File: src/main/java/me/thesourcecode/sourcebot/commands/misc/doc/JenkinsHandler.java
package me.thesourcecode.sourcebot.commands.misc.doc;

import com.overzealous.remark.Remark;
import me.theforbiddenai.jenkinsparserkotlin.Jenkins;
import me.theforbiddenai.jenkinsparserkotlin.entities.*;
import me.thesourcecode.sourcebot.api.utility.SourceColor;
import me.thesourcecode.sourcebot.listener.DocSelectionListener;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.utils.MarkdownSanitizer;
import net.dv8tion.jda.api.utils.MarkdownUtil;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;

import java.lang.reflect.Field;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class JenkinsHandler {

    private final String baseURL;
    private final String iconURL;
    private final String embedTitle;

    private final Jenkins jenkins;
    private final Remark remark;
    private List<String> classList;

    public JenkinsHandler(String docURL, String iconURL, String embedTitle) {
        this.baseURL = docURL.substring(0, docURL.lastIndexOf("/") + 1).trim();

        this.iconURL = iconURL;
        this.embedTitle = embedTitle;

        jenkins = new Jenkins(docURL);
        remark = new Remark();
        getClassList();
    }

    public String getDocURL() {
        return baseURL;
    }

    /**
     * Gets the information from jenkins and gets the embeds
     *
     * @param query The information being looked for
     * @return The embed for the documentation or a selection menu embed
     */
    public MessageEmbed search(User user, String query) {
        try {
            removeUserFromDocCache(user);
            List<Information> foundInformationList = jenkins.search(query);


            if (foundInformationList.size() == 0) return null;

            EmbedBuilder documentationEmbed = new EmbedBuilder()
                    .setColor(SourceColor.BLUE.asColor())
                    .setAuthor(embedTitle, null, iconURL)
                    .setFooter("Ran By: " + user.getAsTag(), user.getEffectiveAvatarUrl());

            if (foundInformationList.size() == 1) {
                Information information = foundInformationList.get(0);

                return createDocumentationEmbed(documentationEmbed, information).build();
            } else {
                DocSelectionListener.selectionStorageCache.put(user, foundInformationList);
                return createSelectionEmbed(documentationEmbed, foundInformationList).build();
            }
        } catch (Exception ex) {
            return null;
        }
    }

    /**
     * Removes the user from the doc cache if they are in it
     *
     * @param user The user being removed
     */
    private void removeUserFromDocCache(User user) {
        ConcurrentHashMap<User, Map.Entry<Message, JenkinsHandler>> docStorageCache = DocSelectionListener.docStorageCache;
        ConcurrentHashMap<User, List<Information>> selectionStorageCache = DocSelectionListener.selectionStorageCache;


        if (docStorageCache.containsKey(user)) {
            docStorageCache.get(user).getKey().delete().queue();

            selectionStorageCache.remove(user);
            docStorageCache.remove(user);
        }
    }

    /**
     * Creates the selection menu embed
     *
     * @param selectionMenu   The embed builder for the selection menu (this contains the color and author already)
     * @param informationList The list of found information objects
     * @return The selection menu
     */
    private EmbedBuilder createSelectionEmbed(EmbedBuilder selectionMenu, List<Information> informationList) {
        selectionMenu.setTitle("Type the id of the option you would like to select in chat:");
        selectionMenu.setFooter("Type cancel to delete this message.");


        informationList.sort(Comparator.comparing(info -> {
            if (info instanceof MethodInformation) {
                MethodInformation methodInfo = (MethodInformation) info;
                return methodInfo.getName().length();
            }
            return info.getName().length();
        }));

        int count = 1;
        for (Information info : informationList) {
            String name = info.getType() + " " + info.getName();

            String optionText = MarkdownUtil.maskedLink(name, info.getUrl());
            selectionMenu.appendDescription("\n\n**" + count + "** - " + optionText);

            count++;
            if (selectionMenu.getDescriptionBuilder().length() >= 712) {
                String format = "\n\nIds not shown above: %d to %d";
                selectionMenu.appendDescription(String.format(format, count, informationList.size() - 1));
                break;
            }

        }

        return selectionMenu;
    }

    /**
     * Creates the documentation embed for the specified information object
     *
     * @param documentationEmbed The documentation embed (this contains the color and author already)
     * @param information        The information object the embed is being created for
     * @return The modified documentation embed which contains all the information from the information object
     */
    public EmbedBuilder createDocumentationEmbed(EmbedBuilder documentationEmbed, Information information) {
        String infoURL = MarkdownSanitizer.escape(information.getUrl());
        String infoName = MarkdownSanitizer.escape(information.getName());
        String infoRawDescription = information.getRawDescription();

        if (information instanceof ClassInformation) {
            ClassInformation classInfo = (ClassInformation) information;

            List<String> nestedClasses = classInfo.getNestedClassList();
            List<String> methodList = classInfo.getMethodList();
            List<String> enumList = classInfo.getEnumList();
            List<String> fieldList = classInfo.getFieldList();

            if (!nestedClasses.isEmpty()) {
                String nestedClassList = formatList(nestedClasses).replace(infoName + ".", "");
                documentationEmbed.addField("Nested Classes:", nestedClassList, false);
            }
            if (!methodList.isEmpty()) {
                methodList = methodList.stream()
                        .map(methodName -> methodName.substring(0, methodName.indexOf("(")).trim())
                        .collect(Collectors.toList());

                documentationEmbed.addField("Methods:", formatList(methodList), false);
            }
            if (!enumList.isEmpty()) {
                documentationEmbed.addField("Enums:", formatList(enumList), false);
            }
            if (!fieldList.isEmpty()) {
                documentationEmbed.addField("Fields:", formatList(fieldList), false);
            }

            // format list
        } else if (information instanceof MethodInformation) {
            MethodInformation methodInfo = (MethodInformation) information;

            infoName = getInfoName(methodInfo.getClassInfo().getName(), methodInfo.getName());
        } else if (information instanceof EnumInformation) {
            EnumInformation enumInfo = (EnumInformation) information;

            infoName = getInfoName(enumInfo.getClassInfo().getName(), enumInfo.getName());
        } else if (information instanceof FieldInformation) {
            FieldInformation fieldInfo = (FieldInformation) information;

            infoName = getInfoName(fieldInfo.getClassInfo().getName(), fieldInfo.getName());
        }

        Element infoDescriptionElement = Jsoup.parse("<div>" + infoRawDescription + "</div>").selectFirst("div");
        String infoDescription = convertHTMLToMarkdown(convertHyperlinks(infoDescriptionElement));

        String infoHyperLink = MarkdownUtil.maskedLink(infoName, infoURL);

        documentationEmbed.setDescription("**__" + infoHyperLink + "__**")
                .appendDescription("\n" + infoDescription);

        if (!(information instanceof ClassInformation)) {
            Map<String, String> rawExtraInfo = information.getRawExtraInformation();

            rawExtraInfo.forEach((key, value) -> {

                Element valueElement = Jsoup.parse("<div>" + value + "</div>").selectFirst("div");
                String convertedValue = convertHTMLToMarkdown(convertHyperlinks(valueElement));

                documentationEmbed.addField(key, convertedValue, false);
            });
        }


        return documentationEmbed;
    }

    /**
     * Converts raw HTML to markdown while conserving hyperlinks
     *
     * @param rawHTML The html being converted to markdown
     * @return The converted html
     */
    private String convertHTMLToMarkdown(String rawHTML) {
        rawHTML = rawHTML.replaceAll("(<code>.*</code> -)", "§$1");
        String markdown = remark.convertFragment(rawHTML.replaceAll("<code>\\[(.*?)]\\((.*?)\\)</code>", "[<code>$1</code>]($2)"))
                .replaceAll("\\\\(\\W)", "$1").replace("§", "\n");

        if (markdown.replaceAll("\\[.*?]\\(.*?\\)", "").length() == markdown.length()) {
            markdown = limit(markdown, 600);
        }
        return markdown;
    }

    /**
     * Converts `a` elements to hyperlinks
     *
     * @param element The element the `a` elements are coming from
     * @return The raw html with the hyperlinks
     */
    private String convertHyperlinks(Element element) {
        String rawHtml = element.html();

        for (Element hyperLinkElement : element.select("a")) {
            String href = hyperLinkElement.attr("href");
            String text = hyperLinkElement.text();

            if (href == null) continue;

            if ((!href.contains("http") && (href.contains("../") || href.contains("#")))) {
                href = href.replace("../", "");
                href = baseURL + href;
            } else if ((!href.contains("http") && href.contains(".html"))) {
                href = href.contains("/") ? href.substring(href.lastIndexOf("/") + 1) : href;
                href = getClassURL(href.replace(".html", ""));
            }

            String hyperlink = MarkdownUtil.maskedLink(MarkdownSanitizer.escape(text), MarkdownSanitizer.escape(href));
            rawHtml = rawHtml.replace(hyperLinkElement.outerHtml(), hyperlink);
            if (convertHTMLToMarkdown(rawHtml).length() >= 900) {
                rawHtml = rawHtml.substring(0, rawHtml.lastIndexOf(hyperlink) - 3) + " ...";
                break;
            }
        }


        return rawHtml;
    }

    /**
     * Formats a given list into a string
     *
     * @param list The list being formatted
     * @return The formatted list in a string
     */
    private String formatList(List<String> list) {
        StringBuilder formattedList = new StringBuilder();

        list.forEach(item -> {
            if (!formattedList.toString().contains(item) && formattedList.length() <= 512) {
                formattedList.append("`").append(item).append("` ");

                if (formattedList.length() >= 512) {
                    formattedList.append("...");
                }
            }
        });
        return formattedList.toString().trim();
    }

    /**
     * Gets the url to class from a string
     *
     * @param className The name of the class
     * @return The url to the class
     */
    private String getClassURL(String className) {
        List<String> urlList = classList.stream()
                .filter(element -> {
                    element = element.substring(element.lastIndexOf("/") + 1).replace(".html", "");
                    return element.equalsIgnoreCase(className);
                })
                .collect(Collectors.toList());

        if (urlList.size() == 0) return null;

        return urlList.get(0);
    }

    /**
     * Formats the class name and info name
     *
     * @param className The name of the class
     * @param infoName  The name of the information object
     * @return The formatted string
     */
    private String getInfoName(String className, String infoName) {
        infoName = infoName.contains("(") ? infoName.substring(0, infoName.indexOf("(")) : infoName;
        return className + "#" + infoName;
    }

    /**
     * Limit the string to a certain number of characters, adding "..." if it was shortened
     *
     * @param value  The string to limitOptions.
     * @param length The length to limitOptions to (as an int).
     * @return The limited string.
     */
    private String limit(String value, int length) {
        StringBuilder buf = new StringBuilder(value);
        if (buf.length() > length) {
            buf.setLength(length);
            buf.append("...");
        }

        return buf.toString();
    }

    /**
     * Gets the element list from the JenkinsParser api
     */
    private void getClassList() {
        try {
            Field field = Jenkins.class.getDeclaredField("classList");
            field.setAccessible(true);
            this.classList = (List<String>) field.get(jenkins);
            field.setAccessible(false);
        } catch (NoSuchFieldException | IllegalAccessException ex) {
            ex.printStackTrace();
        }
    }

}

File: Extensions/Documentation/src/main/kotlin/net/sourcebot/module/documentation/utility/JenkinsHandler.kt
package net.sourcebot.module.documentation.utility

import me.theforbiddenai.jenkinsparserkotlin.Jenkins
import me.theforbiddenai.jenkinsparserkotlin.entities.*
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.utils.MarkdownSanitizer
import net.dv8tion.jda.api.utils.MarkdownUtil
import net.sourcebot.api.response.ErrorResponse
import net.sourcebot.api.response.Response
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import java.util.stream.Collectors

class JenkinsHandler(
    url: String,
    val iconUrl: String,
    val embedTitle: String
) {

    private val jenkins: Jenkins = Jenkins(url)
    private val baseUrl: String by lazy {
        url.substring(0, url.lastIndexOf("/") + 1).trim()
    }

    fun retrieveDocAlert(cmdMessage: Message, user: User, query: String): Response {
        DocSelectorStorage.removeAndDeleteSelector(user)

        try {
            val infoList: List<Information> = jenkins.search(query)

            if (infoList.isEmpty()) {
                return ErrorResponse(user.name, "Unable to find `$query` in the $embedTitle!")
            }

            val docAlert = DocResponse()
            docAlert.setAuthor(embedTitle, null, iconUrl)

            return if (infoList.size == 1) {
                createDocumentationEmbed(docAlert, infoList[0])
            } else {
                val docStorage = DocSelectorStorage(null, cmdMessage, infoList, this)
                DocSelectorStorage.addSelector(user, docStorage)

                createSelectionEmbed(docAlert, infoList)

            }

        } catch (ex: Exception) {
            ex.printStackTrace() //- This is for debug purposes when enabled
            val errDesc = "Unable to find `$query` in the $embedTitle!"
            return ErrorResponse(user.name, errDesc)
        }
    }

    fun createDocumentationEmbed(docAlert: DocResponse, information: Information): DocResponse {
        var infoName: String = MarkdownSanitizer.sanitize(information.name)
        val infoRawDescription: String = information.rawDescription
        val infoUrl: String = MarkdownSanitizer.sanitize(information.url)

        when (information) {
            is ClassInformation -> {

                val nestedClassList: List<String> = information.nestedClassList.stream()
                    .map { it.replace("$infoName.", "") }
                    .collect(Collectors.toList())

                val methodList: List<String> = information.methodList.stream()
                    .map { it.substringBefore("(").trim() }
                    .collect(Collectors.toList())

                docAlert.attemptAddEmbedField(nestedClassList, "Nested Classes:")
                docAlert.attemptAddEmbedField(methodList, "Methods:")
                docAlert.attemptAddEmbedField(information.enumList, "Enums:")
                docAlert.attemptAddEmbedField(information.fieldList, "Fields:")
            }

            is MethodInformation -> {
                val methodName = information.name.substringBefore("(").trim()
                infoName = getInfoName(information.classInfo.name, methodName)
            }

            is FieldInformation -> {
                infoName = getInfoName(information.classInfo.name, information.name)
            }

            is EnumInformation -> {
                infoName = getInfoName(information.classInfo.name, information.name)
            }

            else -> {
                throw Error("Found an object that does not belong.")
            }


        }

        val infoDescElement: Element = Jsoup.parse("<div>$infoRawDescription</div>").selectFirst("div")
        val infoDescription: String = convertHyperlinksToMarkdown(infoDescElement, infoUrl)
            .toMarkdown()
            .truncate(600)

        val infoHyperlink: String = MarkdownUtil.maskedLink(infoName, infoUrl)
        docAlert.setDescription("**__${infoHyperlink}__**\n$infoDescription")

        if (information !is ClassInformation) {
            val rawExtraInfo: Map<String, String> = information.rawExtraInformation

            rawExtraInfo.forEach { (key, value) ->
                val modifiedValue = value.replace("\n", "<br>")


                val valueElement: Element = Jsoup.parse("<div>$modifiedValue</div>").selectFirst("div")
                val convertedValue = convertHyperlinksToMarkdown(valueElement, infoUrl)
                    .toMarkdown()
                    .replace("<\\[`?([^`]+)`?]\\((.*)\\)>".toRegex(), "[`<$1>`]($2)")

                val fieldValue = if (key.equals("Parameters:", true)) {
                    convertedValue.truncate(250)
                } else {
                    if (convertedValue.length > 500) {
                        convertedValue.truncate(500)
                    }
                    convertedValue
                }

                docAlert.addField(key, fieldValue, false)
            }
        }

        return docAlert
    }

    private fun createSelectionEmbed(docAlert: DocResponse, infoList: List<Information>): DocResponse {
        docAlert.setTitle("Type the id of the option you would like to select in chat:")
        docAlert.setFooter("Type cancel to delete this message.")

        infoList.sortedByDescending { it.name }

        for (i in infoList.indices) {

            if (docAlert.descriptionBuilder.length >= 712) {
                val format = "\n\nIds not shown above: %d to %d"
                docAlert.appendDescription(String.format(format, i + 1, infoList.size))
                break
            }

            val info = infoList[i]

            var name = "${info.type} ${info.name}"
            if (info is MethodInformation) {
                val className = info.classInfo.name.substringBefore("<")
                name = "${info.type} $className#${info.name}"
            }

            val optionText = MarkdownUtil.maskedLink(name, info.url)

            docAlert.appendDescription("\n\n**${i + 1}** - $optionText")

        }

        return docAlert
    }

    private fun EmbedBuilder.attemptAddEmbedField(fieldList: List<String>, fieldName: String): EmbedBuilder {
        if (fieldList.isNotEmpty()) {
            this.addField(fieldName, formatList(fieldList), false)
        }

        return this
    }

    private fun formatList(list: List<String>): String {
        val strBuilder = StringBuilder()

        list.stream()
            .filter { !strBuilder.toString().contains(it, true) && strBuilder.length <= 512 }
            .forEach {
                val builderLength = strBuilder.length + it.length

                val itemName = it.substringBefore("<")
                val appendStr = if (builderLength >= 512) "`$itemName`..." else "`$itemName` "
                strBuilder.append(appendStr)
            }

        return strBuilder.toString().trim()
    }

    private fun getInfoName(className: String, infoName: String): String {
        return "$className#${infoName.substringAfter("(" + 1)}"
            .replace("<.*?>+".toRegex(), "")
    }

    private fun convertHyperlinksToMarkdown(element: Element, url: String): String {
        var html: String = element.outerHtml()

        element.select("a").stream()
            .filter { it.attr("href") != null }
            .forEach {

                val baseClassUrl: String = url.substringBeforeLast("#")
                var hrefUrl: String = it.attr("href")
                val text: String = it.html().toMarkdown()

                hrefUrl = with(hrefUrl) {
                    when {
                        contains("http") -> hrefUrl
                        contains(".html") -> {
                            val pkgUrl = url.substringBeforeLast("/")
                            "$pkgUrl/$this"
                        }
                        contains("../") || contains("#") && !contains(".html") -> {

                            if (contains("#") && !contains("/")) {
                                return@with baseClassUrl + url
                            }

                            val modifiedHref = replace("../", "")

                            if (it.hasAttr("title")) {
                                val newPackage = it.attr("title")
                                    .replace("class in ", "")
                                    .replace(".", "/")
                                    .trim()

                                val parentPkgDir = "/${newPackage.substringBefore("/")}/"
                                val basePkgUrl = url.substringBeforeLast(parentPkgDir) + parentPkgDir

                                basePkgUrl + modifiedHref
                            } else {
                                baseUrl + modifiedHref
                            }

                        }
                        contains(".html", true) -> {
                            val modifiedHref = substring(lastIndexOf("/") + 1)
                                .substringBeforeLast(".")

                            retrieveClassUrl(modifiedHref)?.trim() ?: return@forEach
                        }
                        else -> this
                    }

                }

                if (hrefUrl.isNotEmpty()) {
                    val isCodeBlock = it.parent().tagName()?.equals("code", true) ?: false

                    val modifiedText = if (isCodeBlock && !text.contains("`")) "`$text`" else text

                    hrefUrl = MarkdownSanitizer.escape(hrefUrl)
                    val hyperlink: String = MarkdownUtil.maskedLink(modifiedText, hrefUrl)
                        .replace("%29", ")")

                    html = html.replace(it.outerHtml(), hyperlink)
                }

            }


        // Removes code tags that surround a hyperlink
        return html.replace("<code>(\\[.*?\\).*?)</code>".toRegex(), "$1")
    }


    private fun retrieveClassUrl(className: String): String? {
        val urlList = jenkins.classList.stream()
            .filter {
                val modifiedElement = it.substring(it.lastIndexOf("/") + 1).removeSuffix(".html")
                return@filter modifiedElement.equals(className, true)
            }.collect(Collectors.toList())

        return if (urlList.size == 0) null else urlList[0]
    }
}